inventory_directory option for ansible provisioner

fajpunk opened the pull request:
Add an `inventory_directory` setting to the Ansible provisioner that allows a
user to specify a directory in which the Packer Ansible provisioner would write
the generated inventory file.  If a value is specified for this setting, then
have the Packer Ansible provisioner pass this directory as the -i arg when it
calls ansible.

This implements the closed issue https://github.com/hashicorp/packer/issues/4510.

This is a simpler implementation (with tests) of part of https://github.com/mitchellh/packer/pull/4047 .  To repeat a comment on that PR, my use case for this functionality is:

I have an set of Ansible config that I use to provision some infrastructure in
the normal Ansible way. I have an inventory directory with host_vars and
group_vars; this inventory directory is not the same directory that my
playbooks are in. I want to be able to use the variables that I've specified in
this inventory directory when I'm using the Packer Ansible provisioner also.
Currently, I can't do this, because the Packer Ansible provisioner runs Ansible
with -i <gnerated inventory tmp file>.

I can't solve this by passing my own -i <my inventory dir> in the
extra_arguments key of the provisioner, because this overrides the
Packer-generated inventory file, and Ansible then doesn't find the new host.

If I could specify a directory in which the Packer Ansible provisioner would
write the generated inventory file, and have the Packer Ansible provisioner
pass this directory as the -i arg when it calls ansible, then I could use my
existing inventory and it's host_vars and group_vars variables with the Packer
Ansible provisioner.

For my particular use case, I don't care what the name of the generated
inventory file is; I'd imagine an inventory_directory option in the Packer
Ansible provisioner that makes it write the generated inventory file to that
directory and pass that directory to the -i argument when Ansible is called.
This would allow an Ansible playbook used by the Packer Ansible provisioner to
use variables specified in `host_vars` and `group_vars` in this inventory
directory.

Closes #4510

bhcleek commented:
`getInventoryArg` and the line that was changed to use it should be reverted, because if given an inventory directory, Ansible will use all the inventory files in that directory, and Packer should not be triggering its plays to be run against other inventory.

fajpunk commented:
D'oh!  Good call.  I had been testing with hosts specified in my test playbook.  I reverted that change, and also added a check that the inventory directory exists if it is specified.

bhcleek commented on provisioner/ansible/provisioner.go:
```diff
@@ -382,6 +391,21 @@ func validateFileConfig(name string, config string, req bool) error {
 	return nil
 }
 
+func validateDirectoryConfig(name string, config string, req bool) error {
```
There's no need for `config` and `req`. This function is only called when `name` is guaranteed to be something other than the empty string.

fajpunk replied:
I got rid of `req`, but `config` is used in the errors that `validateDirectoryConfig` returns, so I left it in.  Let me know if I should change the error message so that this function doesn't need the `config` param.

bhcleek replied:
`config` never varies and is used in every error string created by `validateDirectoryConfig`. Is there any reason to keep the `config` parameter?

fajpunk replied:
It's there to mirror `validateFileConfig`, which is used to validate the state of different files required by different options.  In those cases, a different value is passed into `config` for each different option.  I suppose the `config` param in `validateDirectoryConfig` falls under YAGNI maybe, since that function is only used once ever (in validating `inventory_directory`.  Given that, I could get rid of `validateDiretoryConfig` entirely and inline the whole thing, I guess; I was trying to make the validation of the `inventory_directory` setting consistent with the validation of the other settings that depended on stuff in the filesystem.

fajpunk replied:
What's the approach you want me to take here?  Should I just get rid of the `validateDirectoryConfig` function, and inline all of that logic in the `Prepare` function?

bhcleek replied:
I would suggest keeping the `validateDirectoryConfig` function, but removing its `config` parameter.

fajpunk replied:
What should I put into the error message?  Should I just hardcode `inventory_directory`?

fajpunk replied:
Ok, I hardcoded `inventory_directory` and changed the name of the function to `validateInventoryDirectoryConfig`.

fajpunk commented:
That CI failure on https://github.com/hashicorp/packer/pull/4760/commits/f45f4568dbfb7ffdad7eed608eb19369353ecf1c
seems unrelated to my changes:

```
--- FAIL: TestPfxRoundTriEcdsa (0.01s)
	pkcs12_test.go:67: pkcs12: error decrypting PKCS#8 shrouded key bag: pkcs12: decryption error, incorrect padding

```
https://travis-ci.org/hashicorp/packer/jobs/219423058#L192

bhcleek commented:
LGTM

korotovsky commented:
Hi, as an author of https://github.com/hashicorp/packer/pull/4047 I really appreciate this PR! It also solves the problem that I had. When it could be merged into `master`? Would be nice, because have a custom build of fork is not that nice :)

korotovsky commented:
@mwhooker Thanks a lot!

korotovsky commented:
Oooops. This PR is not fully covers my PR :) I still have the issue when `ansible` provisioner specifies its own inventory file and I can't override it. Any suggestions? How we could solve it? 

Currently it generates the following command:

```
==> virtualbox-iso: Executing Ansible: ansible-playbook --extra-vars packer_build_name=virtualbox-iso \
packer_builder_type=virtualbox-iso -i project-name/inventory/dev/inventory/packer-provisioner-ansible969995519 \
 ~/project-playbooks/project-name/project-name.yml --inventory-file project-name/inventory/dev \
--private-keyproject-name/packer/keys/insecure_id_rsa -vvv
```

So there is no way to override `-i project-name/inventory/dev/inventory/packer-provisioner-ansible969995519` even if I specified my own inventory as a directly where `host_vars` and `group_vars` are located.

@mwhooker @fajpunk

mwhooker commented:
no. more work may need to be done. This could be as simple as exporting `inventoryFile` in the config. PRs welcome.
